ghmirror/core/mirror_requests.py

```diff
@@ -118,49 +180,46 @@ def online_request(method, url, auth, data=None, url_params=None):
         if last_mod is not None:
             headers['If-Modified-Since'] = last_mod
 
-    resp = requests.request(method=method,
-                            url=url,
-                            headers=headers,
-                            timeout=REQUESTS_TIMEOUT,
-                            params=parameters)
+    resp = _online_request(method=method,
+                           url=url,
+                           headers=headers,
+                           parameters=parameters,
+                           cached_response=cached_response)
 
     if resp.status_code == 304:
-        if len(cached_response.json()) == per_page_elements and \
-           not cached_response.links:
+        return _handle_not_changed(cached_response, per_page_elements,
+                                   headers, method, url, parameters,
+                                   cache, cache_key)
 
-            headers.pop('If-None-Match')
-            resp = requests.request(method=method,
-                                    url=url,
-                                    headers=headers,
-                                    timeout=REQUESTS_TIMEOUT,
-                                    params=parameters)
+    if 'X-Cache' not in resp.headers:
+        LOG.info('ONLINE GET CACHE_MISS %s', url)
+        resp.headers['X-Cache'] = 'ONLINE_MISS'
+    _cache_response(resp, cache, cache_key)
+    return resp
 
-            LOG.info('ONLINE GET CACHE_MISS %s', url)
-            resp.headers['X-Cache'] = 'ONLINE_MISS'
-            _cache_response(resp, cache, cache_key)
-            return resp
 
-        LOG.info('ONLINE GET CACHE_HIT %s', url)
-        cached_response.headers['X-Cache'] = 'ONLINE_HIT'
-        return cached_response
+def _should_error_response_be_served_from_cache(response):
+    """Parse a response to check if we should serve contents
+    from cache
 
-    # When we hit the API limit, let's try to serve from cache
-    if _should_serve_from_cache(resp):
-        if cached_response is None:
-            LOG.info('RATE_LIMITED GET CACHE_MISS %s', url)
-            resp.headers['X-Cache'] = 'RATE_LIMITED_MISS'
-            return resp
-        LOG.info('RATE_LIMITED GET CACHE_HIT %s', url)
-        cached_response.headers['X-Cache'] = 'RATE_LIMITED_HIT'
-        return cached_response
+    :param response: requests module response
+    :type response: requests.Response
 
-    LOG.info('ONLINE GET CACHE_MISS %s', url)
-    resp.headers['X-Cache'] = 'ONLINE_MISS'
-    _cache_response(resp, cache, cache_key)
-    return resp
+    :return: a bool indicating if we should serve from cache or not
+    and the headers that we should return
+    :rtype: bool, string
```

Comment:
it's kinda implicit that you're returning a tuple, bu you can be more explicit here:
```
    :rtype: tuple[bool, string]
```